README.txt
= Implementation =
This extension uses Portable PHP password hashing framework [1]
to create salted hashed passwords. The framework has been properly
tested and reviewed. The framework is used amongst others by
Drupal CMS and Wordpress.
By that, you could import or export user accounts between
these applications without having to change passwords.


= Integration =
 * FE authentication service that works with hashed passwords
 * (JS)/PHP eval functions that allows creation of hashed passwords in BE
 * "forgot password" functionality for sysext felogin
 * "forgot password" functionality for ext feuser_admin
 * create user/edit user functionality in regards to password field for ext feuser_admin


= Extension configuration variables =
 * onlyAuthService
	If enabled, the provided service is the only authentication service for FE users.
	This means when authentication fails with this service, no other one will be tried to use.
 * forcePHPasswd
	If enabled, authentication will only work with passwords generated by provided
	Portable PHP password hashing framework. Neither existing passwords (clear-text, md5)
	nor updated ones (see section "Update existing FE user records") will be used for
	authentication.
 * updatePasswd
	If enabled, FE user passwords will be converted to passwords of
	Portable PHP password hashing framework during authentication.
	(This will not work if forcePHPasswd is activated)


= Update existing FE user records =
Note: If you've been tested previous versions of this extension (version <=0.0.3),
		you will need to convert passwords back to clear-text or md5 ones.
After installing the extension, you are able to update the existing passwords
of FE users so that Portable PHP password hashing framwork could use them.
There's a detection mechanism that tries to find out if an update is needed.
If so, you will find function "UPDATE!"  when extension has been selection
(just to help you: other functions would be "Information", "Edit files", "Backup/Delete")
Due to security reasons, the extension's service won't use these updated passwords
by default. If you would like to use them, disable extension configuration
variable "forcePHPasswd".


The provided template is exactly that one of extension feuser_admin,
but reduced by one single line which allowed an instant login by
transmitting credentials via GET.

Usage with extension feuser_admin:
please set following TypoScript to enable usage of
Portable PHP password hashing framework:

# plugin.feadmin.fe_users.parseValues.password = trim,saltedHash

if not set, passwords will be created the common way (clear-text, md5)
and FE users might not be able to log in with the provided service


Licenses:
This extension is distributed under GPL.

Modified Portable PHP password hashing framework
retrieved from Drupal CMS is distributed under GPL
[1] http://www.openwall.com/phpass

Extension icon is retrieved from Mini set by Mark James
http://www.famfamfam.com/lab/icons/mini/
Distributed under GPL

ChangeLog
2010-02-26  Marcus Krause   <marcus#exp2010@t3sec.info>

	* Security fix: Authentication Bypass in frontend user authentication
	* Feature: backported PHPass library from sysext:saltedpasswords (TYPO3 4.3+)

2009-10-27  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #4629: Update-Module not very clear (thanks to Georg Ringer)
	* Bug #4927: UPDATE! show up only if last created password is not encrypted (thanks to Dimitri Koenig)


2009-09-04  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #4456: Allow overwriting translations via TS. (thanks to Jonas Duebi)
	* Bug #4456: Add XCLASS code to felogin xclass (thanks to Jonas Duebi)
	* Task #4448: Cleanup: Remove randombyte-generator
	* Feature #4446: Add short labels for extension configuration template


2009-05-18  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #3436: Problems with Localization


2009-05-18  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #3365: PHP complains about deprecated call-time pass-by-reference


2009-05-10  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #3254: Code cleanup - remove unnecessary method call (user record)


2009-03-24  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #2956: Forget Password mail not send due to wrong forgothash


2009-03-23  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #2951: Form evaluation fails for activated salted passwords in one TYPO3 mode only


2009-03-07  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #2323: HTML Entities in Plain-Text-Mail
	* Feature #2742: Get template configuration from felogin
	* Bug #2743: felogin "forgot password" form does not work with t3sec_saltedpw
	* Bug #2741: BE users cannot change their own passwords - add warning in extension configuration


2009-01-31  Marcus Krause   <marcus#exp2009@t3sec.info>

	* implement Unit-Test that tests update necessity for decreased standard hash count (follow-up to #2441)


2009-01-10  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #2439: In case of a modifed HASHCOUNT, a non-valid PHPass password would be set
	* Bug #2438: Missing 2nd argument for method call updatePassword in service class
	* Bug #2440: In PHPass method isHashUpdateNeeded() current hashcount is accessed by using class constant
	* Feature #2441: Only require a hash update if hash count has been increased
	* Task #2437: Use shorthand logical comparisons


2009-01-06  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Task #2410: Adjust mail address of maintainer to reflect new year


2008-12-19  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bug #2339: Extension key "t3sec_saltedpw" was NOT loaded
	* Bug #2351: class.ext_update.php throws warning if current password length < 4


2008-12-16  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Task #2324: Improve wording and wording consistency in file ChangeLog
	* Task #2325: Improve readability of extension manual


2008-12-15  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bugfix #2241:  Manual does not get rendered on typo3.org
	* Bugfix #2310:  LL label "ll_enter_your_data" only available in english
	* Bugfix #2311:  ux_felogin_pi1: Send mail with password
	* Bugfix #2312:  $LANG not available
	* Bugfix #2319:  BE-User password is salted, but user auth is used the normal
	* Task #2320:    Comment cleanup in felogin XCLASS


2008-12-13  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bugfix #2241: Manual does not get rendered on typo3.org
	* Bugfix #2243: Refer to forge instead of bugtracker in regards to extension bugs.


2008-12-11  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Feature #2300: Backport improvements introduced by new method in trunk::t3lib_div
	* Feature #2301: Use randomBytes generator to create passwords
	* Feature #2302: Use randomBytes generator of t3lib_div for TYPO3 versions >= 4.3


2008-12-10  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bugfix #2242: Extension title too long


2008-11-27  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Limit felogin XCLASS usage to TYPO3 versions below 4.2.4
	* Bugfix: add bulk processing protection for sysext felogin


2008-11-26  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Feature: extend update functionality to be_users records also
	* Commenting out XCLASS for sysext felogin


2008-11-24  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Add extension manual


2008-11-23  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Bugfix: Install tool was still creating md5 hashed passwords


2008-11-22  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Centralize loginSecurityLevel check
	* Bugfix: access method was not returning a boolean for empty tables
	* Bugfix: ext felogin was always assuming that salted password hashes are used in the FE


2008-11-21  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Add additional default eval und parse functions to TypoScript setup
	* Bugfix in fe_adminLib.inc(CREATE): Salted password hashes were shown to user
	* Make configuration loginSecurityLevel set to 'normal' mandatory


2008-11-21  Michael Stucki  <michael@typo3.org>

	* Bugfix in fe_adminLib.inc(EDIT): Salted password hashes were shown to user
	* Bugfix in ->compareUident(): Plaintext passwords were treated as md5


2008-11-20  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Take a user's lockDomain configuration into account
	* Add service for BE user authentication


2008-11-19  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Set service priority to 60 (> 50 [tx_sv_auth] && < 75 [OpenID])


2008-11-19  Lars Houmark    <lars@typo3.org>

	* Bugfix: only enabled records were converted by update function


2008-11-17  Dmitry Dulepov  <dmitry@typo3.org>

	* Changed "new" to t3lib_div::makeInstance()


2008-11-14  Marcus Krause   <marcus#exp2009@t3sec.info>

	* Initial work on this extension

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext: "t3sec_saltedpw"
#
# Auto generated 26-02-2010 05:58
#
# Manual updates:
# Only the data in the array - anything else is removed by next write.
# "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'TYPO3 Security - Salted user password hashes',
	'description' => 'Uses Portable PHP password hashing framework for storing passwords. Integrates into system extension "felogin" and extension "feuser_admin". Use SSL to secure datatransfer! PLEASE READ MANUAL FIRST!',
	'category' => 'services',
	'shy' => 0,
	'dependencies' => 'cms',
	'conflicts' => 'kb_md5fepw,newloginbox,pt_feauthcryptpw',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'stable',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => 'fe_users',
	'clearCacheOnLoad' => 1,
	'lockType' => '',
	'author' => 'Marcus Krause',
	'author_email' => 'marcus#exp2009@t3sec.info',
	'author_company' => 'TYPO3 Security Team',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'version' => '0.2.13',
	'constraints' => array(
		'depends' => array(
			'typo3' => '4.2.5-4.3.0',
			'php' => '5.1.0-0.0.0',
			'cms' => '',
		),
		'conflicts' => array(
			'kb_md5fepw' => '',
			'newloginbox' => '',
			'pt_feauthcryptpw' => '',
		),
		'suggests' => array(
			'felogin' => '',
			'feuser_admin' => '',
			'phpunit' => '',
		),
	),
	'_md5_values_when_last_written' => 'a:24:{s:9:"ChangeLog";s:4:"2a57";s:11:"LICENSE.txt";s:4:"c635";s:20:"class.ext_update.php";s:4:"7e26";s:21:"ext_conf_template.txt";s:4:"b26e";s:12:"ext_icon.gif";s:4:"4324";s:17:"ext_localconf.php";s:4:"6d8a";s:14:"ext_tables.php";s:4:"bea3";s:14:"ext_tables.sql";s:4:"9fb2";s:28:"ext_typoscript_constants.txt";s:4:"bbc9";s:24:"ext_typoscript_setup.txt";s:4:"57ef";s:48:"tests/class.tx_t3secsaltedpw_phpass_testcase.php";s:4:"e2ac";s:34:"res/ux/class.ux_tx_felogin_pi1.php";s:4:"3dd8";s:30:"res/ux/class.ux_tx_install.php";s:4:"983e";s:25:"res/ux/ux_fe_adminLib.inc";s:4:"9a5a";s:41:"res/lib/class.tx_t3secsaltedpw_phpass.php";s:4:"a863";s:31:"res/tmpl/fe_admin_fe_users.tmpl";s:4:"8ba9";s:30:"res/tmpl/felogin_template.html";s:4:"d51e";s:44:"res/staticlib/class.tx_t3secsaltedpw_div.php";s:4:"7ff2";s:28:"res/LL/felogin_locallang.xml";s:4:"1d96";s:42:"res/eval/class.tx_t3secsaltedpw_salted.php";s:4:"d82d";s:45:"res/eval/class.tx_t3secsaltedpw_salted_be.php";s:4:"30dc";s:45:"res/eval/class.tx_t3secsaltedpw_salted_fe.php";s:4:"58a1";s:14:"doc/manual.sxw";s:4:"3807";s:34:"sv1/class.tx_t3secsaltedpw_sv1.php";s:4:"02a4";}',
	'suggests' => array(
	),
);

?>